// phi-colors/library/source/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var phiColors;
(function (phiColors) {
    phiColors.phi = 1.618033988749895;
    phiColors.HslHMin = -Math.PI;
    phiColors.HslHMax = Math.PI;
    phiColors.HslSMin = 0.0;
    phiColors.HslSMax = 2.0 / 3.0;
    phiColors.HslLMin = 0.0;
    phiColors.HslLMax = 1.0;
    phiColors.rLumaRate = 0.299;
    phiColors.gLumaRate = 0.587;
    phiColors.bLumaRate = 0.114;
    var toHex = function (i) {
        var result = Math.round(255 * i).toString(16).toUpperCase();
        if (1 === result.length) {
            result = "0" + result;
        }
        return result;
    };
    phiColors.rgbForStyle = function (expression) {
        return "#"
            + toHex(expression.r)
            + toHex(expression.g)
            + toHex(expression.b);
    };
    phiColors.rgbaForStyle = function (expression) {
        return phiColors.rgbForStyle(expression) + toHex(expression.a);
    };
    phiColors.rgbFromStyle = function (style) {
        var r = 0.0;
        var g = 0.0;
        var b = 0.0;
        while ("#" === style.substr(0, 1)) {
            style = style.substr(1);
        }
        if (3 === style.length) {
            r = (parseInt(style.substr(0, 1), 16) * 0x11) / 255.0;
            g = (parseInt(style.substr(1, 1), 16) * 0x11) / 255.0;
            b = (parseInt(style.substr(2, 1), 16) * 0x11) / 255.0;
        }
        if (6 === style.length) {
            r = parseInt(style.substr(0, 2), 16) / 255.0;
            g = parseInt(style.substr(2, 2), 16) / 255.0;
            b = parseInt(style.substr(4, 2), 16) / 255.0;
        }
        return { r: r, g: g, b: b };
    };
    phiColors.rgbaFromStyle = function (style) {
        var r = 0.0;
        var g = 0.0;
        var b = 0.0;
        var a = 1.0;
        while ("#" === style.substr(0, 1)) {
            style = style.substr(1);
        }
        if (3 === style.length || 4 === style.length) {
            r = (parseInt(style.substr(0, 1), 16) * 0x11) / 255.0;
            g = (parseInt(style.substr(1, 1), 16) * 0x11) / 255.0;
            b = (parseInt(style.substr(2, 1), 16) * 0x11) / 255.0;
            if (4 === style.length) {
                a = (parseInt(style.substr(3, 1), 16) * 0x11) / 255.0;
            }
        }
        if (6 === style.length || 8 === style.length) {
            r = parseInt(style.substr(0, 2), 16) / 255.0;
            g = parseInt(style.substr(2, 2), 16) / 255.0;
            b = parseInt(style.substr(4, 2), 16) / 255.0;
            if (8 === style.length) {
                a = parseInt(style.substr(6, 2), 16) / 255.0;
            }
        }
        return { r: r, g: g, b: b, a: a };
    };
    phiColors.xyzToLength = function (xyz) { return Math.sqrt(Math.pow(xyz.x, 2) + Math.pow(xyz.y, 2) + Math.pow(xyz.z, 2)); };
    phiColors.rgbToXyz = function (expression) { return ({ x: expression.r, y: expression.g, z: expression.b }); };
    phiColors.rgbToHue = function (expression) {
        var hueXy = {
            x: expression.r - ((expression.g / 2) + (expression.b / 2)),
            y: Math.sqrt(Math.pow(expression.g, 2) - Math.pow(expression.g / 2, 2))
                - Math.sqrt(Math.pow(expression.b, 2) - Math.pow(expression.b / 2, 2))
        };
        return Math.atan2(hueXy.y, hueXy.x);
    };
    phiColors.rgbToLuma = function (expression) { return (expression.r * phiColors.rLumaRate) + (expression.g * phiColors.gLumaRate) + (expression.b * phiColors.bLumaRate); };
    phiColors.rgbToLightness = function (expression) { return (expression.r + expression.g + expression.b) / 3.0; };
    phiColors.calcSaturation = function (expression) {
        var lightness = phiColors.rgbToLightness(expression);
        return phiColors.xyzToLength({ x: expression.r - lightness, y: expression.g - lightness, z: expression.b - lightness });
    };
    phiColors.rgbToSaturation = function (expression) { return phiColors.calcSaturation(expression) * phiColors.calcSaturation({ r: 1.0, g: 0.0, b: 0.0 }); };
    phiColors.rgbToHsl = function (expression) {
        return ({
            h: phiColors.rgbToHue(expression),
            s: phiColors.rgbToSaturation(expression),
            l: phiColors.rgbToLightness(expression)
        });
    };
    phiColors.rgbaToHsla = function (expression) {
        return ({
            h: phiColors.rgbToHue(expression),
            s: phiColors.rgbToSaturation(expression),
            l: phiColors.rgbToLightness(expression),
            a: expression.a,
        });
    };
    phiColors.hslToRgbElement = function (expression, Angle) { return expression.l + expression.s * Math.cos(expression.h - (Math.PI * 2) / 3.0 * Angle); };
    phiColors.hslToRgb = function (expression) {
        return ({
            r: phiColors.hslToRgbElement(expression, 0.0),
            g: phiColors.hslToRgbElement(expression, 1.0),
            b: phiColors.hslToRgbElement(expression, 2.0)
        });
    };
    phiColors.hslaToRgba = function (expression) {
        return ({
            r: phiColors.hslToRgbElement(expression, 0.0),
            g: phiColors.hslToRgbElement(expression, 1.0),
            b: phiColors.hslToRgbElement(expression, 2.0),
            a: expression.a,
        });
    };
    phiColors.regulateHue = function (expression) {
        var h = expression.h;
        while (h < -Math.PI) {
            h += Math.PI * 2;
        }
        while (Math.PI < h) {
            h -= Math.PI * 2;
        }
        var result = {
            h: h,
            s: expression.s,
            l: expression.l,
        };
        return result;
    };
    phiColors.clipLightness = function (expression) {
        return ({
            h: expression.h,
            s: expression.s,
            l: Math.max(0.0, Math.min(1.0, expression.l)),
        });
    };
    phiColors.clipSaturation = function (expression) {
        var rgb = phiColors.hslToRgb(expression);
        var overRate = Math.max((rgb.r < 0.0) ? (expression.l - rgb.r) / expression.l :
            (1.0 < rgb.r) ? (rgb.r - expression.l) / (1.0 - expression.l) :
                1.0, (rgb.g < 0.0) ? (expression.l - rgb.g) / expression.l :
            (1.0 < rgb.g) ? (rgb.g - expression.l) / (1.0 - expression.l) :
                1.0, (rgb.b < 0.0) ? (expression.l - rgb.b) / expression.l :
            (1.0 < rgb.b) ? (rgb.b - expression.l) / (1.0 - expression.l) :
                1.0);
        var result = {
            h: expression.h,
            s: expression.s / overRate,
            l: expression.l,
        };
        return result;
    };
    phiColors.regulateHsl = function (expression) { return phiColors.clipSaturation(phiColors.clipLightness(phiColors.regulateHue(expression))); };
    phiColors.regulateHsla = function (expression) {
        var result = phiColors.clipSaturation(phiColors.clipLightness(phiColors.regulateHue(expression)));
        return {
            h: result.h,
            s: result.s,
            l: result.l,
            a: expression.a,
        };
    };
    phiColors.clipRgb = function (expression) {
        return ({
            r: Math.max(0.0, Math.min(1.0, expression.r)),
            g: Math.max(0.0, Math.min(1.0, expression.g)),
            b: Math.max(0.0, Math.min(1.0, expression.b)),
        });
    };
    phiColors.clipRgba = function (expression) {
        return ({
            r: Math.max(0.0, Math.min(1.0, expression.r)),
            g: Math.max(0.0, Math.min(1.0, expression.g)),
            b: Math.max(0.0, Math.min(1.0, expression.b)),
            a: Math.max(0.0, Math.min(1.0, expression.a)),
        });
    };
    phiColors.generate = function (base, h, s, l, a, isAlignLuma) {
        if (isAlignLuma === void 0) { isAlignLuma = true; }
        var hsla = {
            h: base.h,
            s: base.s,
            l: base.l,
            a: base.a,
        };
        if (undefined !== h && 0.0 !== h) {
            hsla.h += Math.PI * 2 / phiColors.phi * h;
        }
        if (undefined !== s && 0.0 !== s) {
            hsla.s = s < 0.0 ?
                hsla.s / Math.pow(phiColors.phi, -s) :
                phiColors.HslSMax - ((phiColors.HslSMax - hsla.s) / Math.pow(phiColors.phi, s));
        }
        if (undefined !== l && 0.0 !== l) {
            hsla.l = l < 0.0 ?
                hsla.l / Math.pow(phiColors.phi, -l) :
                1.0 - ((1.0 - hsla.l) / Math.pow(phiColors.phi, l));
        }
        if (undefined !== a && 0.0 !== a) {
            hsla.a = a < 0.0 ?
                hsla.a / Math.pow(phiColors.phi, -a) :
                1.0 - ((1.0 - hsla.a) / Math.pow(phiColors.phi, a));
        }
        if (isAlignLuma) {
            var baseLuuma = phiColors.rgbToLuma(phiColors.hslToRgb({ h: base.h, s: base.s, l: hsla.l }));
            var luuma = phiColors.rgbToLuma(phiColors.hslToRgb(hsla));
            hsla.l += baseLuuma - luuma;
        }
        return phiColors.regulateHsla(hsla);
    };
})(phiColors = exports.phiColors || (exports.phiColors = {}));
